test: countable (switchable, ownable) components

NueloSE opened the pull request:
**Issue(s):** #70

### Description

Please provide a brief description of the changes made in this pull request and how they address the related issue.

### Checklist

- [x] **CI Verifier:** Run `./scripts/cairo_programs_verifier.sh` successfully
- [x] **Contract Tests:** Added tests to cover the changes

NueloSE commented:
@julio4  ready for review

NueloSE commented:
@julio4 ready for review

julio4 reviewed (changes requested):
LGTM, thank you!
Could you:
- Add test cases to verify that events are emitted?
- Wrap contracts/components inside anchor comments and edit the corresponding markdown files in `src` to use `rustdoc_include ...:anchor` to hide by default tests in the book

julio4 commented:
@NueloSE Do you need any help to finish this?

NueloSE commented:
Yes I need help I haven't been able to figure how to add test to test emitted events for switchable and ownable

julio4 commented:
> @julio4 Yes I need help I haven't been able to figure how to add test to test emitted events for switchable and ownable

Read the tests of [this section](https://starknet-by-example.voyager.online/getting-started/basics/events.html?highlight=starknet::testing::pop_log). (Be sure to toggle hidden lines)

NueloSE commented:
Hi @julio4 
- I have added the ANCHOR tags
- Implemented the test for emitted events in the reference sample but the test keeps failing for the emitted events. 
A screenshot has been attached of the error
![Screenshot from 2024-06-19 10-43-37](https://github.com/NethermindEth/StarknetByExample/assets/124416278/0cfd0047-2e33-4f52-9ebf-73559080425c)

misicnenad commented on listings/applications/components/src/countable.cairo:
```diff
@@ -24,3 +27,68 @@ pub mod countable_component {
 }
+// ANCHOR_END: countable_component
+
+// ANCHOR: MockContract
```
```suggestion
```
The chapter will not be displaying `MockContract`.

misicnenad commented on listings/applications/components/src/countable.cairo:
```diff
@@ -24,3 +27,68 @@ pub mod countable_component {
+    #[abi(embed_v0)]
+    impl CountableImpl = countable_component::Countable<ContractState>;
+}
+// ANCHOR_END: MockContract
```
```suggestion
```

misicnenad commented on listings/applications/components/src/countable.cairo:
```diff
@@ -24,3 +27,68 @@ pub mod countable_component {
+    }
+
+    #[test]
+    fn test_multiply_increment() {
```
```suggestion
    fn test_multiple_increments() {
```
nit: the name wrongly implies the counter is being "multiplied", changing it to "multiple_increments" states the test is incrementing multiple times.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -1,11 +1,16 @@
 use starknet::ContractAddress;
 
 #[starknet::interface]
+// ANCHOR: interface
```
Same as before, both the interface and the component should be wrapped within the same anchor, as they are displayed together (see chapter [Ownable](https://starknet-by-example.voyager.online/components/ownable.html#ownable)).
E.g. you can just use a single anchor `// ANCHOR: component`.

misicnenad commented on listings/applications/components/src/countable.cairo:
```diff
@@ -24,3 +27,68 @@ pub mod countable_component {
+#[cfg(test)]
+mod test {
+    use super::MockContract;
+    use components::countable::{ICountableDispatcher, ICountableDispatcherTrait};
```
```suggestion
    use super::{ICountableDispatcher, ICountableDispatcherTrait};
```

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -24,19 +29,19 @@ pub mod ownable_component {
-    struct OwnershipTransferredEvent {
-        previous: ContractAddress,
-        new: ContractAddress
+    #[derive(Copy, Drop, Debug, PartialEq, starknet::Event)]
```
You don't need to derive the `Copy` trait for any of these events.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+    use super::MockContract;
+    use super::ownable_component::{Event, OwnershipRenouncedEvent, OwnershipTransferredEvent};
+    use starknet::ContractAddress;
+    use components::ownable::{IOwnableDispatcher, IOwnableDispatcherTrait};
```
```suggestion
    use super::{IOwnableDispatcher, IOwnableDispatcherTrait};
```

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+            Option::Some(Event::OwnershipTransferredEvent(OwnershipTransferredEvent { previous:contract_address, new:owner}))
+        );
+    }
+}
```
Missing new line character

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+    fn deploy() -> (IOwnableDispatcher, ContractAddress) {
+        let caller = contract_address_const::<'caller'>();
+        let (contract_address, _) = deploy_syscall(
+            MockContract::TEST_CLASS_HASH.try_into().unwrap(), caller.into(), array![].span(), false
```
```suggestion
            MockContract::TEST_CLASS_HASH.try_into().unwrap(), 0, array![].span(), false
```
unnecessary to set the salt.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+        set_contract_address(Zero::zero());
+        let (ownable, _) = deploy();
+
+        ownable.transfer_ownership(contract_address_const::<'new_owner'>());
```
```suggestion
```
The deployment will panic during constructor call, the test will never even get to this line.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+        set_contract_address(contract_address);
+        let (ownable, address) = deploy();
+        ownable.renounce_ownership();
+        set_contract_address(address);
```
```suggestion
```
unnecessary

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+        let contract_address = contract_address_const::<'owner'>();
+        set_contract_address(contract_address);
+        let (ownable, address) = deploy();
+        let owner = contract_address_const::<'owner'>();
```
```suggestion
        let new_owner = contract_address_const::<'new_owner'>();
```
'owner' is the original owner, so the test is transferring ownership from an owner to themselves, which would make little sense.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+        let (ownable, address) = deploy();
+        let owner = contract_address_const::<'owner'>();
+        ownable.transfer_ownership(owner);
+        set_contract_address(address);
```
```suggestion
```
unnecessary

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -14,11 +14,11 @@ pub mod switchable_component {
 
-    #[derive(Drop, starknet::Event)]
-    struct SwitchEvent {}
+    #[derive(Copy, Drop, Debug, PartialEq, starknet::Event)]
```
```suggestion
    #[derive(Drop, Debug, PartialEq, starknet::Event)]
```
Don't need the `Copy` trait

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -48,4 +48,86 @@ pub mod switchable_component {
+mod MockContract {
+    use super::switchable_component;
+
+    component!(path: switchable_component, storage: switchable, event: switchableEvent);
```
```suggestion
    component!(path: switchable_component, storage: switchable, event: SwitchableEvent);
```

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -48,4 +48,86 @@ pub mod switchable_component {
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        switchableEvent: switchable_component::Event
```
```suggestion
        SwitchableEvent: switchable_component::Event
```

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -48,4 +48,86 @@ pub mod switchable_component {
 }
 // ANCHOR_END: component
 
+// ANCHOR: MockContract
```
```suggestion
```
will never display the mock contract in chapters.

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -48,4 +48,86 @@ pub mod switchable_component {
+mod test {
+    use super::MockContract;
+    use super::switchable_component::{Event, SwitchEvent};
+    use components::switchable::{ISwitchableDispatcher, ISwitchableDispatcherTrait};
```
```suggestion
    use super::{ISwitchableDispatcher, ISwitchableDispatcherTrait};
```

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -48,4 +48,86 @@ pub mod switchable_component {
+
+    #[test]
+    fn test_switch_event() {
+        let contract_address = contract_address_const::<'address'>();
```
```suggestion
```
not needed

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -48,4 +48,86 @@ pub mod switchable_component {
+            Option::Some(Event::SwitchEvent(SwitchEvent { }))
+        );
+    }
+}
```
Missing new line

misicnenad commented:
> Hi @julio4
> 
> * I have added the ANCHOR tags
> * Implemented the test for emitted events in the reference sample but the test keeps failing for the emitted events.
>   A screenshot has been attached of the error
>   ![Screenshot from 2024-06-19 10-43-37](https://private-user-images.githubusercontent.com/124416278/341017662-0cfd0047-2e33-4f52-9ebf-73559080425c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NTgwNjIsIm5iZiI6MTcxODk1Nzc2MiwicGF0aCI6Ii8xMjQ0MTYyNzgvMzQxMDE3NjYyLTBjZmQwMDQ3LTJlMzMtNGY1Mi05ZWJmLTczNTU5MDgwNDI1Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMVQwODE2MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTc5OTY1OWNjZTE4NTBlMmI4ZmU2NGQ4MzYyMTYyY2I4YWFlNzM0ZTU3MGQ5N2Q5YzJkMWVlNDJkMDk5NTFiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KrLHKLBvw9WnBQ2CZHoCl5jM4Q9rwnTh28h587TI3KY)

@julio4 I tested this locally and I experienced the same issues. It's like events are not being emitted from components.

@NueloSE could you try updating your tests to use Foundry (see [Testing Smart Contracts](https://book.cairo-lang.org/ch17-02-testing-smart-contracts.html))? Maybe these are some issues that they solved.

julio4 commented:
> I tested this out with Foundry too, and found that there's a bug when spying on events emitted from components. Created a ticket for this to be investigated further, see [starkware-libs/cairo#5861](https://github.com/starkware-libs/cairo/issues/5861).

Good find, we can just comment the event spying for now.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +98,159 @@ pub mod ownable_component {
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        OwnableEvent: ownable_component::Event,
```
```suggestion
        #[flat]
        OwnableEvent: ownable_component::Event,
```
@julio4 @NueloSE I [found out](https://github.com/starkware-libs/cairo/issues/5861#issuecomment-2186267348) that adding the `#[flag]` attribute above the component's event (in the contract that imports it) fixes the [event spying issues](https://github.com/NethermindEth/StarknetByExample/pull/205#issuecomment-2182249492).
We don't have to [comment out the event asserts](https://github.com/NethermindEth/StarknetByExample/pull/205#issuecomment-2186217896).

misicnenad replied:
Ignore this comment, we can include event asserts without using the `#[flat]` attribute

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +96,146 @@ pub mod ownable_component {
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
```
```suggestion
    #[derive(Drop, Debug, PartialEq, starknet::Event)]
```
Part of event test fixes

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +96,146 @@ pub mod ownable_component {
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
```
```suggestion
    pub enum Event {
```

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +96,146 @@ pub mod ownable_component {
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
```
```suggestion
```
I know I was the one who suggested you include this attribute, but I actually found out how to make the tests pass *without* it (see https://github.com/NethermindEth/StarknetByExample/pull/205/files#r1657310788).

The issue with `#[flat]` is that you don't _always want to flatten your event structure, as it might cause name conflicts_. So we need a way to test the event emissions without flattening them.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +96,146 @@ pub mod ownable_component {
+        ownable.renounce_ownership();
+        assert_eq!(
+            starknet::testing::pop_log(address),
+            Option::Some(Event::OwnershipRenouncedEvent(OwnershipRenouncedEvent { previous:contract_address }))
```
```suggestion
            Option::Some(MockContract::Event::OwnableEvent(OwnershipRenouncedEvent { previous:contract_address }.into()))
```
Explainer: the issue is that you're expecting the component event type rather than the contract event type. Remember, `MockContract` emits the `Event` type of `MockContract`, whose first variant is itself an enum, specifically the `Event` enum of the component. The test now passes.

misicnenad replied:
This test can also be merged with `test_renounce_ownership`, since they are testing the same flow.

misicnenad commented on listings/applications/components/src/ownable.cairo:
```diff
@@ -93,3 +96,146 @@ pub mod ownable_component {
+        ownable.transfer_ownership(new_owner);
+        assert_eq!(
+            starknet::testing::pop_log(address),
+            Option::Some(Event::OwnershipTransferredEvent(OwnershipTransferredEvent { previous:contract_address, new:new_owner}))
```
```suggestion
            Option::Some(MockContract::Event::OwnableEvent(OwnershipTransferredEvent { previous:contract_address, new:new_owner}.into()))
```

misicnenad replied:
This test can also be merged with `test_transfer_ownership_by_owner`, since they are testing the same flow.

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -49,3 +49,85 @@ pub mod switchable_component {
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
```
```suggestion
    #[derive(Drop, Debug, PartialEq, starknet::Event)]
```

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -49,3 +49,85 @@ pub mod switchable_component {
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
```
```suggestion
    pub enum Event {
```

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -49,3 +49,85 @@ pub mod switchable_component {
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
```
```suggestion
```

misicnenad commented on listings/applications/components/src/switchable.cairo:
```diff
@@ -49,3 +49,85 @@ pub mod switchable_component {
+        switchable.switch();
+        assert_eq!(
+            starknet::testing::pop_log(contract_address),
+            Option::Some(Event::SwitchEvent(SwitchEvent { }))
```
```suggestion
            Option::Some(MockContract::Event::SwitchableEvent(SwitchEvent { }.into()))
```

misicnenad replied:
This test can also be merged with `test_switch`, since they are testing the same flow.

misicnenad reviewed (changes requested):
Changes required to make the original test suite pass (without using `#[flat]`)

NueloSE commented:
Ready for Review

misicnenad reviewed (approved):
Beautiful!

Let's just wait until @julio4 gives it one final look, and that's it :rocket:

julio4 reviewed (approved):
Thank you for adding test for countable and updating a little bit the switchable and ownable tests.
I cleaned up some files, feel free to check and I'll merge if everything is ok!

NueloSE commented:
lgtm
